File: examples/python/output/list_tree.py
#!/usr/bin/env python3
"""Nested lists and tree views.

    make run-example EX=python/output/list_tree
"""

from __future__ import annotations

import sparcli as sc


def print_lists() -> None:
    """Markers, nesting and a styled marker."""
    steps = sc.List(sc.ListOpts(marker=sc.ListMarker.NUMBER,
                                marker_style=sc.Style.bold(sc.Color.CYAN)))
    steps.add('Check out the repository')
    build = steps.add('Build the project')
    steps.add('Run the test suite')

    # A sub-list is owned by the parent item.
    sub = build.sub(sc.ListOpts(marker=sc.ListMarker.ALPHA_LOWER, indent=2))
    sub.add('make')
    sub.add('make qa')
    steps.print()
    sc.println('')

    notes = sc.List(sc.ListOpts(marker=sc.ListMarker.BULLET, bullet='→'))
    notes.add('Long items are word-wrapped and continuation lines align '
              'under the first word of the item.')
    notes.add(sc.Text.from_markup('Items can be [bold green]rich text[/].'))
    notes.print()


def print_tree() -> None:
    """A file-system-like tree; add() returns a node usable as a parent."""
    tree = sc.Tree(sc.TreeOpts(type=sc.BorderType.SINGLE,
                               connector_color=sc.Color.CYAN))
    dir_style = sc.Style.bold(sc.Color.BLUE)

    root = tree.add('sparcli/', style=dir_style)
    src = tree.add('src/', root, style=dir_style)
    tree.add('core/', src, style=dir_style)
    tree.add('output/', src, style=dir_style)
    docs = tree.add('docs/', root, style=dir_style)
    tree.add('examples.md', docs, prefix='▪ ',
             prefix_style=sc.Style(fg=sc.Color.GREEN))
    tree.add('Makefile', root)
    tree.print()


def main() -> None:
    print_lists()
    sc.println('')
    print_tree()


if __name__ == '__main__':
    main()

File: examples/rust/output/list_tree.rs
//! Nested lists and tree views.
//!
//!   cargo run -p sparcli --example output_list_tree

use sparcli::*;

fn main() {
    print_lists();
    println("", Style::default());
    print_tree();
}

/// Markers, nesting and a styled marker.
fn print_lists() {
    let mut steps = List::new(ListOpts::new().marker(ListMarker::Number));
    steps.add("Check out the repository", Style::default());
    let build = steps.add("Build the project", Style::default());
    steps.add("Run the test suite", Style::default());

    // A sub-list is owned by the parent item.
    let mut sub = build.sub(ListOpts::new().marker(ListMarker::AlphaLower).indent(2));
    sub.add("make", Style::default());
    sub.add("make qa", Style::default());
    steps.print();
    println("", Style::default());

    let mut notes = List::new(ListOpts::new().marker(ListMarker::Bullet).bullet("→"));
    notes.add(
        "Long items are word-wrapped and continuation lines align under \
         the first word of the item.",
        Style::default(),
    );
    notes.print();
}

/// A file-system-like tree; `add` returns a node usable as a parent.
fn print_tree() {
    let mut tree = Tree::new(TreeOpts::new());
    let root = tree.add("sparcli/", TreeNode::root(), Style::bold().fg(Color::BLUE));
    let src = tree.add("src/", root, Style::bold().fg(Color::BLUE));
    tree.add("core/", src, Style::default());
    tree.add("output/", src, Style::default());
    let docs = tree.add("docs/", root, Style::bold().fg(Color::BLUE));
    tree.add("examples.md", docs, Style::new().fg(Color::GREEN));
    tree.add("Makefile", root, Style::default());
    tree.print();
}
